### PR TITLE
RDS IAM support

### PR DESCRIPTION
**Pull request by pjdufour-truss**
## Description

Enabling IAM Authentication for connecting to RDS.  Since we got migrations using the same database connection code, our migrations, scheduled tasks, and services should all be able to use this new logic. 

Will need to test thoroughly.

Tests:
- [ ] Scheduled Tasks on Experimental Pass!
- [ ] Migrations on Experimental Pass!
- [ ] Service on Experimental Works!
- [ ] Acceptance tests don't break (we can't actually run acceptance tests for this, since they are run outside the VPC).

## Reviewer Notes

Local Migrations

```shell
make db_dev_destroy && make db_dev_start && DB_TIMEOUT=30 make db_dev_create && make db_dev_migrate
```

Experimental Migrations

```shell
make db_deployed_migrations_destroy && make db_deployed_migrations_start && DB_TIMEOUT=30 make db_deployed_migrations_create && AWS_ASSUME_ROLE_DURATION=45m make run_experimental_migrations
```

**Questions**

- Do we want to use a manifest?  Or continue to run all migrations in folder?
- Is there a better way to manage the migration "path" (MIGRATION_PATH env var) that points to folders containing migrations?
- How long do migrations take on your local machine now?  (local dev and deployed migrations).

## Setup

None

## Code Review Verification Steps

* [ ] Code follows the guidelines for [Logging](https://github.com/transcom/mymove/tree/master/docs/backend.md#logging)
* [ ] The requirements listed in
 [Querying the Database Safely](https://github.com/transcom/mymove/tree/master/docs/backend.md#querying-the-database-safely)
 have been satisfied.
* Any new migrations/schema changes:
  * [ ] Follow our guidelines for zero-downtime deploys (see [Zero-Downtime Deploys](https://github.com/transcom/mymove/tree/master/docs/database.md#zero-downtime-migrations))
  * [ ] Have been communicated to #dp3-engineering
  * [ ] Secure migrations have been tested using `scripts/run-prod-migrations`
* [ ] There are no aXe warnings for UI.
* [ ] This works in [Supported Browsers and their phone views](https://github.com/transcom/mymove/tree/master/docs/adr/0016-Browser-Support.md) (Chrome, Firefox, IE, Edge).
* Any new client dependencies (Google Analytics, hosted libraries, CDNs, etc) have been:
  * [ ] Communicated to @ntwyman
  * [ ] Added to the list of [network dependencies](https://github.com/transcom/mymove#client-network-dependencies)
* [ ] Tested in the Experimental environment (for changes to containers, app startup, or connection to data stores)
* [ ] User facing changes have been reviewed by design.
* [ ] Request review from a member of a different team.
* [ ] Have the Pivotal acceptance criteria been met for this change?

## References

* [Pivotal story](https://www.pivotaltracker.com/story/show/161565546) for this change
* [Pivotal story](https://www.pivotaltracker.com/story/show/161566038) for this change

## Screenshots

N.A.

### CODE REVIEW

File: config/app-client-tls.container-definition.json [hunk 1]
```diff
@@ -16,16 +9,24 @@
     "exec",
     "app-{{ .environment }}",
     "--",
-    "/bin/milmove",
-    "serve"
+    "/bin/milmove"
```
Comment (chrisgilmerproj) [inline config/app-client-tls.container-definition.json]: I was talking to Nick about the security of the client containers and the difference between `entryPoint` and `command`. The `entryPoint` cannot be changed by someone who has access to the system running the container but the `command` does. So for security we probably ought to lock down the `entryPoint` to use only the command we want that container to run. That's a long winded way for me to say we probably ought to leave `"serve"` in the `entryPoint` section instead of moving it. We also ought to correct the other container definitions.

Comment (pjdufour-truss) [reply config/app-client-tls.container-definition.json]: Hm... I like that idea.  Add a little more friction.  Maybe we'd want to move all the CMD part for the container definitions into entrypoint actually...  I also found this blog post useful.

https://aws.amazon.com/blogs/opensource/demystifying-entrypoint-cmd-docker/

File: Dockerfile [hunk 2]
```diff
@@ -1,15 +1,17 @@
...
 
 COPY config /config
+
 COPY swagger/* /swagger/
+
 COPY build /build
 
 ENTRYPOINT ["/bin/milmove"]
 
-CMD ["serve", "--debug-logging"]
```
Comment (chrisgilmerproj) [inline Dockerfile]: Should this be:

```dockerfile
ENTRYPOINT ["/bin/milmove", "serve"]
CMD ["--debug-logging"]
```

I'm wondering how I'd run this docker file locally.

File: Dockerfile.migrations [hunk 3]
```diff
@@ -1,22 +1,15 @@
...
 
-# Install tools needed in container
-RUN apk update
-RUN apk add postgresql-client python3 ca-certificates bash --no-cache
-RUN pip3 install awscli
+WORKDIR /
 
-WORKDIR /scripts
 ENTRYPOINT ["/bin/milmove"]
-CMD ["migrate", "-p", "/migrate/migrations"]
```
Comment (chrisgilmerproj) [inline Dockerfile.migrations]: Same question here, if this is

```dockerfile
ENTRYPOINT ["/bin/milmove", "migrate"]
CMD ["-p", "/migrate/migrations"]
```

Then I can run this locally, otherwise I feel like I have to rely on the invocation in a `Makefile` or some test code or the ECS task definition?

File: Makefile [hunk 4]
```diff
@@ -1,9 +1,9 @@
 NAME = ppp
 DB_NAME_DEV = dev_db
-DB_NAME_PROD_MIGRATIONS = prod_migrations
+DB_NAME_DEPLOYED_MIGRATIONS = deployed_migrations
```
Comment (chrisgilmerproj) [inline Makefile]: I think this PR would be easier to understand if we pulled out this renaming into its own PR. I can approve that PR today too because this is a good change.

File: Makefile [hunk 5]
```diff
@@ -437,8 +448,8 @@ server_test_coverage: server_deps server_generate mocks_generate db_test_reset d
 db_dev_destroy: ## Destroy Dev DB
 ifndef CIRCLECI
 	@echo "Destroying the ${DB_DOCKER_CONTAINER_DEV} docker database container..."
-	docker rm -f $(DB_DOCKER_CONTAINER_DEV) || \
-		echo "No database container"
+	docker rm -f $(DB_DOCKER_CONTAINER_DEV) || echo "No database container"
+	rm -fr mnt/db_dev # delete mount directory if exists
```
Comment (chrisgilmerproj) [inline Makefile]: I think it would make sense to have this in the `clean:` target as well. Maybe just add `rm -fr mnt/` inside `clean:`.

File: Makefile [hunk 6]
```diff
@@ -451,18 +462,15 @@ endif
...
+			-e POSTGRES_PASSWORD=$(PGPASSWORD) \
 			-p $(DB_PORT_DEV):$(DB_PORT_DOCKER)\
 			$(DB_DOCKER_CONTAINER_IMAGE)
 
 .PHONY: db_dev_create
 db_dev_create: ## Create Dev DB
 	@echo "Create the ${DB_NAME_DEV} database..."
-	DB_NAME=postgres scripts/wait-for-db && \
-		createdb -p $(DB_PORT_DEV) -h localhost -U postgres $(DB_NAME_DEV) || true
+	DB_NAME=postgres scripts/wait-for-db && DB_NAME=postgres psql-wrapper "CREATE DATABASE $(DB_NAME_DEV);" || true
```
Comment (chrisgilmerproj) [inline Makefile]: is this supposed to be `DB_USER=postgres` instead of `DB_NAME`?

File: Makefile [hunk 7]
```diff
@@ -843,6 +848,11 @@ make_test: ## Test make targets not checked by CircleCI
 # ----- START RANDOM TARGETS -----
 #
 
+.PHONY: prune
+prune:  ## Prune docker containers and images
+	docker container prune
+	docker image prune -a
```
Comment (chrisgilmerproj) [inline Makefile]: I'm pretty happy to have this target. I'm wondering if we should go further and do `docker system prune --volumes` which should just do everything?

Btw, several people would find this handy today in case you wanted to make it a separate PR.

File: scripts/run-prod-migrations [hunk 8]
```diff
@@ -1,61 +0,0 @@
...
-aws s3 ls "${SECURE_MIGRATION_BUCKET_NAME}/secure-migrations" > /dev/null
-
-#
-# Run migrations
-#
-
-echo
-echo -e "\e[33mUsing ${SECURE_MIGRATION_BUCKET_NAME} to gather secure migrations\e[39m"
-echo
-proceed "Running production migrations against the ${DB_NAME} database. This will delete everything in that db."
```
Comment (chrisgilmerproj) [inline scripts/run-prod-migrations]: I'm pretty happy to be rid of this script. But I'd like to retain this little piece of it which asks if a user really wants to continue. I could probably be persuaded not to keep it but I think some folks rely on it checking with them.  Maybe ask in #dp3-engineering to get a feel for it?

File: cmd/generate-test-data/main.go [hunk 9]
```diff
@@ -117,8 +120,28 @@ func main() {
...
+	var dbCreds *credentials.Credentials
+	if session != nil {
+		// We want to get the credentials from the session,
+		// because the session conflates the environment, shared, and container metdata config
+		dbCreds = stscreds.NewCredentials(session, v.GetString(cli.DbIamRoleFlag))
+	}
+
 	// Create a connection to the DB
-	dbConnection, err := cli.InitDatabase(v, logger)
+	dbConnection, err := cli.InitDatabase(v, dbCreds, logger)
```
Comment (chrisgilmerproj) [inline cmd/generate-test-data/main.go]: I think here we could just set `dbCreds := nil` because we don't generate test data anywhere that would need the `DbIamFlag`.

File: cmd/clean-migration/main.go [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+	"github.com/transcom/mymove/pkg/migrate"
+)
+
+func main() {
+
+	root := cobra.Command{
+		Use:                   "clean-migration INPUT_FILE|- [OUTPUT_FILE]",
+		DisableFlagsInUseLine: true,
+		Short:                 "Clean migration file",
+		Long:                  "Clean migration file",
```
Comment (chrisgilmerproj) [inline cmd/clean-migration/main.go]: Could you expand the long description a bit?

File: cmd/milmove/migrate.go [hunk 11]
```diff
@@ -96,31 +120,131 @@ func migrateFunction(cmd *cobra.Command, args []string) error {
...
 	if err != nil {
-		logger.Fatal("invalid configuration", zap.Error(err))
+		return errors.Wrap(err, "invalid configuration")
 	}
 
 	if v.GetBool(cli.DbDebugFlag) {
 		pop.Debug = true
 	}
 
+	migrationPath := expandPaths(strings.Split(v.GetString(cli.MigrationPathFlag), ";"))
```
Comment (chrisgilmerproj) [inline cmd/milmove/migrate.go]: `s/migrationPath/migrationPaths/g` - It's a nitpick but plural would be nice since this is a slice.

File: cmd/milmove/migrate.go [hunk 12]
```diff
@@ -96,31 +120,131 @@ func migrateFunction(cmd *cobra.Command, args []string) error {
...
 	}
 
-	migrationPath := v.GetString(cli.MigrationPathFlag)
-	logger.Info(fmt.Sprintf("using migration path %q", migrationPath))
-	mig, err := pop.NewFileMigrator(migrationPath, dbConnection)
+	migrationTableName := dbConnection.MigrationTableName()
+	logger.Info(fmt.Sprintf("tracking migrations using table %q", migrationTableName))
+
+	migrationManifest := expandPath(v.GetString(cli.MigrationManifestFlag))
+	logger.Info(fmt.Sprintf("using migration manifest %q", migrationManifest))
```
Comment (chrisgilmerproj) [inline cmd/milmove/migrate.go]: I think I don't understand what the manifest is for. Can you elaborate for me?

File: cmd/split-migration/main.go [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+	"github.com/transcom/mymove/pkg/migrate"
+)
+
+func main() {
+
+	root := cobra.Command{
+		Use:                   "split-migration INPUT_FILE|- [OUTPUT_FILE]",
+		DisableFlagsInUseLine: true,
+		Short:                 "Split migration file",
+		Long:                  "Split migration file",
```
Comment (chrisgilmerproj) [inline cmd/split-migration/main.go]: Please expand the long definition here.

File: cmd/generate-test-data/main.go (hunk 9 above)
Comment (pjdufour-truss) [reply cmd/generate-test-data/main.go]: ok

File: pkg/migrate/CleanMigration.go [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+	"io"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+var cleanMigrationPattern = regexp.MustCompile("^(\\s*)(COPY)(\\s+)([a-zA-Z._]+)(\\s*)\\((.+)\\)(\\s+)(FROM)(\\s+)(stdin)(;)(\\s*)$")
+
+// CleanMigraton cleans migrations so they can be used in a go connection directly.
+func CleanMigraton(r io.Reader, w chan string) {
```
Comment (chrisgilmerproj) [inline pkg/migrate/CleanMigration.go]:
```suggestion
func CleanMigration(r io.Reader, w chan string) {
```

File: cmd/clean-migration/main.go [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+				if err != nil {
+					return errors.Wrap(err, fmt.Sprintf("error reading from file %q", args[0]))
+				}
+				inputReader = f
+				defer f.Close()
+			}
+
+			out := make(chan string, 1000)
+
+			go migrate.CleanMigraton(inputReader, out)
```
Comment (chrisgilmerproj) [inline cmd/clean-migration/main.go]:
```suggestion
			go migrate.CleanMigration(inputReader, out)
```

File: cmd/clean-migration/main.go [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+
+	"github.com/transcom/mymove/pkg/migrate"
+)
+
+func main() {
+
+	root := cobra.Command{
+		Use:                   "clean-migration INPUT_FILE|- [OUTPUT_FILE]",
```
Comment (chrisgilmerproj) [inline cmd/clean-migration/main.go]: I tried to run `clean-migration migrations/20190410152949_add_new_tdls.up.sql ./out.txt` but I only got stdout output whereas I expected from this that it would go to a file. Did I do it wrong?

File: cmd/clean-migration/main.go [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+				inputReader = f
+				defer f.Close()
+			}
+
+			out := make(chan string, 1000)
+
+			go migrate.CleanMigraton(inputReader, out)
+
+			for line := range out {
+				_, err := fmt.Fprintln(os.Stdout, line)
```
Comment (chrisgilmerproj) [inline cmd/clean-migration/main.go]: Here is where I see it printing to stdout but not to a file? Maybe `os.Stdout` is a default if no `outfile` name is given?

File: cmd/milmove/migrate.go (hunk 11 above)
Comment (chrisgilmerproj) [inline cmd/milmove/migrate.go]:
```suggestion
	migrationPaths := expandPaths(strings.Split(v.GetString(cli.MigrationPathFlag), ";"))
```

File: cmd/milmove/migrate.go [hunk 18]
```diff
@@ -96,31 +120,131 @@ func migrateFunction(cmd *cobra.Command, args []string) error {
...
+	for _, p := range migrationPath {
+		if strings.HasPrefix(p, "s3://") {
+			s3Migrations = true
+			break
+		}
+	}
+
+	var session *awssession.Session
+	if v.GetBool(cli.DbIamFlag) || s3Migrations {
+		c, errorConfig := cli.GetAWSConfig(v, v.GetBool(cli.VerboseFlag))
```
Comment (chrisgilmerproj) [inline cmd/milmove/migrate.go]: Can we use full names instead of single character names?

```suggestion
		config, errorConfig := cli.GetAWSConfig(v, v.GetBool(cli.VerboseFlag))
```

File: cmd/milmove/migrate.go [hunk 19]
```diff
@@ -96,31 +120,131 @@ func migrateFunction(cmd *cobra.Command, args []string) error {
...
+		if errorConfig != nil {
+			return errors.Wrap(errorConfig, "error creating aws config")
+		}
+		s, errorSession := awssession.NewSession(c)
```
Comment (chrisgilmerproj) [inline cmd/milmove/migrate.go]: Same here, I'd rather use a name.

File: pkg/migrate/ListFiles.go [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+import (
+	"os"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/s3"
+)
+
+// ListFiles lists the files in a given directory.
+func ListFiles(p string, s3Client *s3.S3) ([]string, error) {
```
Comment (chrisgilmerproj) [inline pkg/migrate/ListFiles.go]:
```suggestion
func ListFiles(migrationPath string, s3Client *s3.S3) ([]string, error) {
```

File: pkg/migrate/ListFiles.go [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+	if strings.HasPrefix(p, "file://") {
+		f, err := os.Open(p[len("file://"):])
+		if err != nil {
+			return []string{}, err
+		}
+		return f.Readdirnames(-1)
+	} else if strings.HasPrefix(p, "s3://") {
+		parts := strings.SplitN(p[len("s3://"):], "/", 2)
+		if len(parts) == 2 {
```
Comment (chrisgilmerproj) [inline pkg/migrate/ListFiles.go]: Should we throw an error if it doesn't equal 2? However, I think that its likely impossible so do we want to just skip this `if` statement?

File: cmd/milmove/migrate.go [hunk 22]
```diff
@@ -96,31 +120,131 @@ func migrateFunction(cmd *cobra.Command, args []string) error {
...
+
+	migrationManifest := expandPath(v.GetString(cli.MigrationManifestFlag))
+	logger.Info(fmt.Sprintf("using migration manifest %q", migrationManifest))
+
+	var s3Client *s3.S3
+	if s3Migrations {
+		s3Client = s3.New(session)
+	}
+
+	migrationFiles := map[string][]string{}
```
Comment (chrisgilmerproj) [inline cmd/milmove/migrate.go]: This might sound weird but this seems like a great debugging tool: `list-migration-files` (or something). Because it would be nice to pass in a set of migration paths and see the whole set listed out.

File: pkg/migrate/Exec.go [hunk 23]
```diff
@@ -0,0 +1,223 @@
...
+	go ReadInSQL(inputReader, in, true, true, true) // read in lines as a separate thread
+
+	quoted := 0
+	blocks := NewStack()
+	var stmt strings.Builder
+
+	i := 0
+	for {
+
+		//fmt.Println("Blocks:", blocks)
```
Comment (chrisgilmerproj) [inline pkg/migrate/Exec.go]:
```suggestion
```

File: pkg/migrate/ReadInSQL.go [hunk 24]
```diff
@@ -0,0 +1,33 @@
...
+// If dropBlankLines is true, then drops all blank lines.
+func ReadInSQL(in io.Reader, out *Buffer, dropComments bool, dropBlankLines bool, dropSearchPath bool) {
+	scanner := bufio.NewScanner(in)
+	for scanner.Scan() {
+		line := scanner.Text()
+		if strings.HasPrefix(line, "--") {
+			if dropComments {
+				continue
+			}
+		} else if strings.Contains(line, "pg_catalog.set_config('search_path'") {
```
Comment (chrisgilmerproj) [inline pkg/migrate/ReadInSQL.go]: can you add a comment about why this is called out explicitly?

File: pkg/migrate/Exec.go [hunk 25]
```diff
@@ -0,0 +1,223 @@
...
+	"strings"
+	"time"
+)
+
+import (
+	"github.com/gobuffalo/pop"
+	"github.com/pkg/errors"
+)
+
+func Exec(inputReader io.Reader, tx *pop.Connection, wait time.Duration) error {
```
Comment (chrisgilmerproj) [inline pkg/migrate/Exec.go]: Add a comment here about what this does.

```suggestion
// Exec executes ...
func Exec(inputReader io.Reader, tx *pop.Connection, wait time.Duration) error {
```

File: pkg/migrate/Builder.go [hunk 26]
```diff
@@ -0,0 +1,132 @@
...
+				Key:    aws.String(key),
+			})
+			if errGetObject != nil {
+				if aerr, ok := errGetObject.(awserr.Error); ok {
+					fmt.Println(fmt.Sprintf("AWS Error Code: %q, %q, %q", aerr.Code(), aerr.Message(), aerr.OrigErr()))
+				}
+				return errors.Wrap(errGetObject, fmt.Sprintf("error reading migration file at %q", m.Path))
+			}
+
+			switch m.Type {
```
Comment (chrisgilmerproj) [inline pkg/migrate/Builder.go]: The code in this switch statement feels repeated. Could it be pulled out to a function?

File: cmd/split-migration/main.go [hunk 27]
```diff
@@ -0,0 +1,83 @@
...
+			go func() {
+				scanner := bufio.NewScanner(inputReader)
+				for scanner.Scan() {
+					lines <- scanner.Text()
+				}
+				close(lines)
+			}()
+
+			statements := make(chan string, 1000)
+			go migrate.SplitStatements(lines, statements)
```
Comment (chrisgilmerproj) [inline cmd/split-migration/main.go]: Is this where the slowdown happens for running prod migrations? I'm curious if I can find an example to use to play around with. What did you use?

File: cmd/split-migration/main.go [hunk 28]
```diff
@@ -0,0 +1,83 @@
...
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+
+	"github.com/transcom/mymove/pkg/migrate"
+)
+
+func main() {
+
+	root := cobra.Command{
+		Use:                   "split-migration INPUT_FILE|- [OUTPUT_FILE]",
```
Comment (chrisgilmerproj) [inline cmd/split-migration/main.go]: I expected this to go to an outfile but I don't see it.

File: migrations/20180424010930_test_secure_migrations.up.fizz [hunk 29]
```diff
@@ -1 +0,0 @@
-exec("./apply-secure-migration.sh 20180424010930_test.sql")
```
Comment (chrisgilmerproj) [inline migrations/20180424010930_test_secure_migrations.up.fizz]: These files aren't needed anymore because we have a manifest file and because we have the ability to list files from buckets, is that right?

File: pkg/migrate/Exec.go (hunk 23 above)
Comment (chrisgilmerproj) [reply pkg/migrate/Exec.go]: There are a few of these around. Can you strip out the comments before merging?

File: scripts/dump-migrations [hunk 30]
```diff
@@ -0,0 +1,9 @@
+#! /usr/bin/env bash
+#
+#   Show all used packges in our codebase.
```
Comment (chrisgilmerproj) [inline scripts/dump-migrations]: Update the description here.

File: pkg/migrate/prepareCopyFromStdin_test.go [hunk 31]
```diff
@@ -0,0 +1,33 @@
...
+import (
+	"github.com/gobuffalo/pop"
+)
+
+func (suite *MigrateSuite) TestPrepareCopyFromStdinWithSchema() {
+	tableName := "public.transportation_service_provider_performances"
+	listOfColumns := "id, performance_period_start, performance_period_end, traffic_distribution_list_id, quality_band, offer_count, best_value_score, transportation_service_provider_id, created_at, updated_at, rate_cycle_start, rate_cycle_end, linehaul_rate, sit_rate"
+	columns := parseColumns(listOfColumns)
+
+	suite.DB().Transaction(func(tx *pop.Connection) error {
```
Comment (chrisgilmerproj) [inline pkg/migrate/prepareCopyFromStdin_test.go]: Return the error to the transaction so it can be tested there. Otherwise we don't test the transaction bit.
